### PR TITLE
feat: xkeeper metadata implemented

### PR DESCRIPTION
**Pull request by ashitakah**
XKeeper metadata:

- Contract and interface
- Unit test

### CODE REVIEW

File: solidity/test/unit/XKeeperMetadata.t.sol [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    xKeeperMetadata = new XKeeperMetadata();
+    owner = makeAddr('Owner');
+  }
+}
+
+contract UnitXKeeperMetadataSetAutomationVaultMetadata is XKeeperMetadataUnitTest {
+  modifier happyPath(IAutomationVault _automationVault, string calldata _description) {
+    vm.mockCall(address(_automationVault), abi.encodeWithSelector(IAutomationVault.owner.selector), abi.encode(owner));
+    vm.startPrank(owner);
+    _;
```
Comment (0xJabberwock) [inline solidity/test/unit/XKeeperMetadata.t.sol:31]: Why to input `_description` in the `happyPath`?

File: solidity/test/unit/XKeeperMetadata.t.sol [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  function testRevertInlyAutomationVaultOwner(
```
Comment (0xJabberwock) [inline solidity/test/unit/XKeeperMetadata.t.sol:34]: I guess `Inly` should be `Only`. 😆

**Review by 0xJabberwock — DISMISSED** [review]
Minor suggestions below. Looks good to me!

File: solidity/test/unit/XKeeperMetadata.t.sol (hunk 2 above)
Comment (ashitakah) [reply solidity/test/unit/XKeeperMetadata.t.sol:34]: Yep haha!

File: solidity/test/unit/XKeeperMetadata.t.sol (hunk 1 above)
Comment (ashitakah) [reply solidity/test/unit/XKeeperMetadata.t.sol:31]: description has been deleted, as you said is unused

File: solidity/interfaces/IXKeeperMetadata.sol [hunk 3]
```diff
@@ -11,9 +11,10 @@ interface IXKeeperMetadata {
   /**
    * @notice Emitted when the description of an automation vault is set
```
Comment (0xJabberwock) [inline solidity/interfaces/IXKeeperMetadata.sol:12]: `@notice Emitted when the metadata of an automation vault is set`

File: solidity/interfaces/IXKeeperMetadata.sol [hunk 4]
```diff
@@ -24,19 +25,34 @@ interface IXKeeperMetadata {
...
+    string name;
+    string description;
+  }
+
   /*///////////////////////////////////////////////////////////////
                           VIEW FUNCTIONS
   //////////////////////////////////////////////////////////////*/
 
   /**
    * @notice Returns the description of the automation vault
```
Comment (0xJabberwock) [inline solidity/interfaces/IXKeeperMetadata.sol:47]: `@notice Returns the metadata of the automation vault`

File: solidity/interfaces/IXKeeperMetadata.sol [hunk 5]
```diff
@@ -45,7 +61,10 @@ interface IXKeeperMetadata {
   /**
    * @notice Sets the description of the automation vault
```
Comment (0xJabberwock) [inline solidity/interfaces/IXKeeperMetadata.sol:62]: `@notice Sets the metadata of the automation vault`

File: solidity/interfaces/IXKeeperMetadata.sol [hunk 6]
```diff
@@ -24,19 +25,34 @@ interface IXKeeperMetadata {
...
   /*///////////////////////////////////////////////////////////////
                           VIEW FUNCTIONS
   //////////////////////////////////////////////////////////////*/
 
   /**
    * @notice Returns the description of the automation vault
    * @param  _automationVault The automation vault
-   * @return _description The description of the automation vault
+   * @return _name The description and the name of the automation vault
+   * @return _description The description and the name of the automation vault
```
Comment (0xJabberwock) [inline solidity/interfaces/IXKeeperMetadata.sol:50]:
```
@return _name The name of the automation vault
@return _description The description of the automation vault
```

File: solidity/contracts/XKeeperMetadata.sol [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import {IAutomationVault} from '@interfaces/IAutomationVault.sol';
+
+/**
+ * @title  XKeeperMetadata
+ * @notice This contract is used for managing the metadata of automation vaults
+ */
+
+contract XKeeperMetadata is IXKeeperMetadata {
+  /// @inheritdoc IXKeeperMetadata
+  mapping(IAutomationVault _automationVault => IXKeeperMetadata.AutomationVaultMetadata _automationVaultMetadata) public
```
Comment (0xGorilla) [inline solidity/contracts/XKeeperMetadata.sol:14]: As discussed, we need a method which lets us fetch multiple at once:
```
function automationVaultsMetadata(IAutomationVault _automationVault) view external returns (IXKeeperMetadata.AutomationVaultMetadata[] _metadata)
```

**Review by 0xJabberwock — COMMENTED** [review]
👀

File: solidity/contracts/XKeeperMetadata.sol (hunk 7 above)
Comment (ashitakah) [reply solidity/contracts/XKeeperMetadata.sol:14]: 👍

File: solidity/interfaces/IXKeeperMetadata.sol (hunk 3 above)
Comment (ashitakah) [reply solidity/interfaces/IXKeeperMetadata.sol:12]: 👍

File: solidity/interfaces/IXKeeperMetadata.sol (hunk 4 above)
Comment (ashitakah) [reply solidity/interfaces/IXKeeperMetadata.sol:47]: 👍

File: solidity/interfaces/IXKeeperMetadata.sol (hunk 5 above)
Comment (ashitakah) [reply solidity/interfaces/IXKeeperMetadata.sol:62]: 👍

File: solidity/interfaces/IXKeeperMetadata.sol (hunk 6 above)
Comment (ashitakah) [reply solidity/interfaces/IXKeeperMetadata.sol:50]: 👍

File: solidity/test/unit/XKeeperMetadata.t.sol [hunk 8]
```diff
@@ -3,7 +3,24 @@ pragma solidity 0.8.19;
...
 import {Test} from 'forge-std/Test.sol';
 
-import {IXKeeperMetadata, XKeeperMetadata, IAutomationVault} from '@contracts/XKeeperMetadata.sol';
+import {IXKeeperMetadata, XKeeperMetadata, IAutomationVault, EnumerableSet} from '@contracts/XKeeperMetadata.sol';
+
+contract XKeeperMetadataForTest is XKeeperMetadata {
+  using EnumerableSet for EnumerableSet.AddressSet;
+
+  // This is needed because foundry fuzz some values which are repeated
+  EnumerableSet.AddressSet internal _cleanAutomationVaults;
```
Comment (0xJabberwock) [inline solidity/test/unit/XKeeperMetadata.t.sol:12]: `_cleanAutomationVaults` could be deleted.

**Review by 0xJabberwock — APPROVED** [review]

File: solidity/contracts/XKeeperMetadata.sol [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+  /**
+   * @notice List of automationVaults that have metadata
+   */
+  EnumerableSet.AddressSet internal _automationVaults;
+
+  /// @inheritdoc IXKeeperMetadata
+  function automationVaults() external view returns (address[] memory __automationVaults) {
+    __automationVaults = _automationVaults.values();
+  }
```
Comment (0xGorilla) [inline solidity/contracts/XKeeperMetadata.sol:28]: why do we need this and why do we need the list?
Also, NEVER EVER do a loop like this without pagination. Nevertheless, I don't think this function is necessary

File: solidity/test/unit/XKeeperMetadata.t.sol (hunk 8 above)
Comment (ashitakah) [reply solidity/test/unit/XKeeperMetadata.t.sol:12]: 👍

**Review by 0xGorilla — APPROVED** [review]

File: solidity/contracts/XKeeperMetadata.sol [hunk 10]
```diff
@@ -58,8 +48,8 @@ contract XKeeperMetadata is IXKeeperMetadata {
 
     // Add the automation vault to the list of automation vaults and set the metadata
     automationVaultMetadata[_automationVault] = _automationVaultMetadata;
-    _automationVaults.add(address(_automationVault));
```
Comment (0xJabberwock) [inline solidity/contracts/XKeeperMetadata.sol]: `// Add the automation vault to the list of automation vaults` doesn't happen anymore.

Comment (ashitakah) [reply solidity/contracts/XKeeperMetadata.sol]: nice one!

**Review by 0xJabberwock — APPROVED** [review]

**Review by 0xGorilla — APPROVED** [review]

**Pull request merged**